source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -32,8 +33,12 @@ CacheFilter::CacheFilter(const envoy::extensions::filters::http::cache::v3alpha:
     : time_source_(time_source), cache_(http_cache) {}
 
 void CacheFilter::onDestroy() {
-  lookup_ = nullptr;
-  insert_ = nullptr;
+  if (lookup_) {
```

Comment:
I haven't tracked this PR, but no, this is not guaranteed. Can you describe a bit more what is going on or do you want me to review the entire PR?

Reply:
Most of CacheFilter's plugin cache implementations will complete requests asynchronously, and there's no guarantee of what thread those completions will run on, so their callbacks get posted to our dispatcher. Suppose a plugin posts a callback, then (perhaps due to a client disconnect) CacheFilter::onDestroy is called. That callback must happen before the CacheFilter gets deleted (or else it will access a deleted CacheFilter). AFAICT, the current code does give us the correct order (because it runs pending callbacks before it runs deferred deletion), and I'll be surprised if it ever needs to change, but we need to make sure.

I think that all we need to do is add documentation and tests to lock in the current behavior of running pending callbacks before deferred deletion.

If we can't rely on this, we'll have to go back to a prior iteration of the original CacheFilter PR, and use either weak_ptrs or shared_ptrs to the CacheFilter, but that complicates the code and adds bus-locked operations.

Reply:
The solution here is to not use post, but use a "timer" which can be cancelled. Can we switch to that?

Reply:
Sorry nevermind, I didn't read the cross-thread thing. So, yes, right now the problem with post is it can't be cancelled. This has come up multiple times, and is also related to https://github.com/envoyproxy/envoy/issues/12364#issuecomment-674582595 which I'm about to fix because it has bitten us multiple times. I think other than the weak_ptr/shared_ptr dance, the other option is to build this into post itself and make post cancellable. This would require infra work before your change though.

Reply:
> f we can't rely on this, we'll have to go back to a prior iteration of the original CacheFilter PR, and use either weak_ptrs or shared_ptrs to the CacheFilter, but that complicates the code and adds bus-locked operations.

I don't believe this can be relied on in the current implementation.

Reply:
After taking another look at the current implementation, I think I see what you were alluding to. If someone calls deferredDelete (perhaps on something unrelated to us), then the call to post happens, then CacheFilter::onDestroy; then the deletion could happen before the post gets run, and we're in trouble.

We could easily guarantee that callbacks posted before deferredDelete(foo) is called are run before foo is deleted.  The simplest way I see is to have deferred_delete_cb_'s lambda call runPostCallbacks() before it calls clearDeferredDeleteList(). Would such a change would be useful outside of CacheFilter?

Reply:
Suggestion: rather than typing the asynchronous completion of a caching operation to a filter's lifetime, can we tie it to the config or something else? It seems that holding onto all the state of a filter is not necessary to complete (say) a cache insert.

Of course an http cache lookup should be tied to a filter as that's only in service of a request.

Reply:
> We could easily guarantee that callbacks posted before deferredDelete(foo) is called are run before foo is deleted. The simplest way I see is to have deferred_delete_cb_'s lambda call runPostCallbacks() before it calls clearDeferredDeleteList(). Would such a change would be useful outside of CacheFilter?

It's something we can think about, though I'm hesitant to build more assumptions like this into the dispatcher code as it becomes fragile to maintain. We would definitely want to discuss that in a separate PR. I would honestly be more inclined to just implement a cancellable post. I think it's a cleaner interface, would fix your problem, and has come up multiple times before.

> Suggestion: rather than typing the asynchronous completion of a caching operation to a filter's lifetime, can we tie it to the config or something else? It seems that holding onto all the state of a filter is not necessary to complete (say) a cache insert.

Yeah this is another possible design where you have a cache manager that the filter defers to for this sort of thing.

Reply:
@mattklein123 I am following the discussion at #12364. Can you elaborate how is it related to our problem here?

Reply:
More thoughts: if the client closes the connection during an outstanding lookup, we would need to cancel the cache lookup, and note in its state that the request-related data structures referenced in the lookup context are defunct and can no longer be accessed. That way you can allow the request to be retired and destroyed, even if there is an outstanding async cache lookup.

I think this is pretty do-able, having done similar flows in the past for other projects.

Reply:
@jmarantz the problem arises if the cache lookup has already finished and posted the callback to the dispatcher, but then the client closes the connection and the filter chain is destroyed. Currently, there's no guarantee that the posted callback will run before the filter is deleted. If the filter is deleted first, the posted callback will run on a destroyed filter.
One solution here is to capture a weak_ptr to the CacheFilter in the posted callback, and check that the CacheFilter is still alive before accessing any of its members/methods.

Reply:
Can we arrange a destroy-notify callback from the filter chain? The cache system could then mark the filter structure as stale and avoid referencing it further.

Reply:
of course weak_ptr does the job too, if the thing being pointed to is already a shared_ptr.

Reply:
one more thing: the advantage of a callback over a weak_ptr is that you might be able to actively cancel an outstanding lookup request (if that is possible).

Reply:
> @mattklein123 I am following the discussion at #12364. Can you elaborate how is it related to our problem here?

It's loosely related in the sense that we have TLS post/callbacks that are being running beyond the acceptable lifetime.

> one more thing: the advantage of a callback over a weak_ptr is that you might be able to actively cancel an outstanding lookup request (if that is possible).

This is how a cancellable post would be accomplished within dispatcher. This is what I recommend if you want a core solution, otherwise you can implement this yourself.